Multi sig example (#251)
casper-network/docs: source/docs/casper/dapp-dev-guide/tutorials/multi-sig/example.md

# Setting up a Multi-sig Account

In this section, we will explore the commands to setup an account with associated keys and restrict access for deployment and key management to specific keys within the account. 

In this example, we will set up an account with the following weights and thresholds:
- Three associated keys (Manager, Supervisor, and Clerk)
- Weights of the associated keys:
    - Manager - 4
    - Supervisor - 2
    - Clerk - 1
- Weights of the thresholds:
    - Key management - 4
    - Deploy - 3

:::note

For all the commands used in this tutorial, here are a few pointers:
- The `node-address` attribute is a combination of a peer node IP address from the Mainnet or Testnet prefixed with `7777`, which is the RPC port. 
- The `chain-name` attribute, is `casper-test` for Testnet and `casper` for Mainnet.
- The `amount` attribute is the cost of the transaction in motes.

> - Hexadecimal public key is found in the `public_key_hex` file, copy the contents and paste it in the command that calls for it.

Author: Updated to "The hexadecimal public key is found in the public_key_hex file. Copy the contents of this file and paste it in the command that calls for it."

:::

> ## Creating Accounts

Reviewer: I am not sure if we need to add the links manually such as {#creating-accounts}?

Author: My understanding is it is not needed unless we are hyperlinking to the topic within the document.

Create three sets of keys one each for the Manager, Supervisor and Clerk. 

To create accounts on the Casper blockchain, you need to generate keys and fund your accounts, for more information see [Account and Cryptographic Keys](../../keys.md). To fund your Testnet account, see [Funding Testnet Accounts](../../../workflow/testnet-faucet.md).

## Deploying the Keys Manager Contract

Before you deploy the keys manager contract, you must clone the keys-manager repo and build the contract, see [Contract Example](contract.md) for more information.

You can deploy the keys manager contract using the `put-deploy` command as illustrated below.

```bash
casper-client put-deploy \
--chain-name casper-test \
--node-address http://138.201.54.44:7777 \
--secret-key <path to secret_key.pem> \
--session-path <path to keys-manager.wasm> \
--payment-amount 10000000000
```

<details>
<summary>Sample output with deploy hash</summary>

{
  "id": -4915929071409710835,
  "jsonrpc": "2.0",
  "result": {
    "api_version": "1.4.4",
    "deploy_hash": "3d55c71ae0892c9f5b63be56c8ca3e107e39e812ebc925383881cfa003aaced7"
  }
}

</details>

> ### Check deploy status {#check-deploy-status}

Reviewer: ```suggestion
### Checking Deploy Status {#check-deploy-status}
```

Author: This is a level 3 heading and hence not init capitalized or starts with a gerund.

                  "transform": "Identity"
                },
                {
                  "key": "balance-98d945f5324f865243b7c02c0417ab6eac361c5c56602fd42ced834a1ba201b6",
                  "transform": "Identity"
                },
                {
                  "key": "balance-34b99d631b7545a56af663801f8b86c0f38cc014b002e9f7e0c0b2a19e789f14",
                  "transform": "Identity"
                },
                {
                  "key": "balance-98d945f5324f865243b7c02c0417ab6eac361c5c56602fd42ced834a1ba201b6",
                  "transform": {
                    "WriteCLValue": {
                      "bytes": "00",
                      "cl_type": "U512",
                      "parsed": "0"
                    }
                  }
                },
                {
                  "key": "balance-34b99d631b7545a56af663801f8b86c0f38cc014b002e9f7e0c0b2a19e789f14",
                  "transform": {
                    "AddUInt512": "100000000000"
                  }
                }
              ]
            },
            "transfers": []
          }
        }
      }
    ]
  }
}

</details>

> ### View account details {#view-account-details}

Reviewer: ```suggestion
### Viewing Account Details {#view-account-details}
```

Author: This is a level 3 heading and hence not init capitalized or starts with a gerund.

The following command gets the account details after the deploy is successful:

```
casper-client get-account-info \
--public-key <hexadecimal public key> \
--node-address http://138.201.54.44:7777
```

> In the output of this command, you can see the key weight, key management threshold, and deploy threshold. Also observe the main purse structure, it has the `keys_manager_hash` uref address. This uref address is used to find the smart contract's [session hash](#find-session-hash).

Reviewer: ```suggestion
In the output of this command, you can see the key weight, key management threshold, and deploy threshold. Also, observe the main purse structure, which has the `keys_manager_hash` uref address. This uref address is used to find the smart contract's [session hash](#find-session-hash).
```

{
  "id": 3805776763924596530,
  "jsonrpc": "2.0",
  "result": {
    "account": {
      "account_hash": "account-hash-694c000a8282d039ace20fbad7a692f4d800a7fd6db08848bc95f8a9f7e4273",
      "action_thresholds": {
        "deployment": 1,
        "key_management": 1
      },
      "associated_keys": [
        {
          "account_hash": "account-hash-694c000a08282d039ace20fbad7a692f4d800a7fd6db08848bc95f8a9f7e4273",
          "weight": 1
        }
      ],
      "main_purse": "uref-250554daef44c5274f726f2556b62da7dbd6d997d6a93dea925104529db12e6a-007",
      "named_keys": [
        {
          "key": "hash-fc8eaf47329d45c76c4f80ec4e6fdb08e74dbe4ae0391447c5ebd15f1c962476",
          "name": "keys_manager"
        },
        {
          "key": "uref-3141039f23ae092028da3b8222ebb5ae5d8409e745b507903e5f7600fa42aada-007",
          "name": "keys_manager_hash"
        }
      ]
    },
    "api_version": "1.4.4",
    "merkle_proof": "[26588 hex chars]"
  }
}

</details>

In the above output, you can see the named keys for the contract and the uref for keys_manager_hash. This uref hash will be used to find the session hash of the contract.

> ### Find session hash

Reviewer: ```suggestion
### Finding the Session Hash
```

Author: This is a level 3 heading and hence not init capitalized or starts with a gerund.

  }
}

You can use this state root hash in the next section to find the session hash of the smart contract. 

</details>

**Uref hash of the keys manager contract**

You can find the uref hash in the account information of the account used to deploy the keys manager contract, see [View Account Details](#view-account-details-view-account-details). 

**Session hash for keys manager contract**

Use the following command to get the session hash of the smart contract: 

```bash
casper-client query-state \
--node-address http://138.201.54.44:7777 \
--key <uref of keys_manager_hash> \
--state-root-hash <network state root hash>
```

<details>
<summary>Sample output with session hash</summary>

{
  "id": -3744898570362638107,
  "jsonrpc": "2.0",
  "result": {
    "api_version": "1.4.4",
    "block_header": null,
    "merkle_proof": "[34692 hex chars]",
    "stored_value": {
      "CLValue": {
        "bytes": "fc8eaf47329d45c76c4f80ec4e6fdb08e74dbe4ae0391447c5ebd15f1c962476",
        "cl_type": {
          "ByteArray": 32
        },
        "parsed": "fc8eaf47329d45c76c4f80ec4e6fdb08e74dbe4ae0391447c5ebd15f1c962476"
      }
    }
  }
}

</details>

> ## Set Manager's Key Weight to 4

Reviewer: ```suggestion
## Setting the Manager's Key Weight to 4
```

<details>
<summary>Sample output with one associated key</summary>

{
  "id": -244324861334765287,
  "jsonrpc": "2.0",
  "result": {
    "account": {
      "account_hash": "account-hash-694c109a08282d039ace20fbad7a692f4d800a7fd6db08848bc95f8a9f7e4273",
      "action_thresholds": {
        "deployment": 1,
        "key_management": 1
      },
      "associated_keys": [
        {
          "account_hash": "account-hash-694c109a08282d039ace20fbad7a692f4d800a7fd6db08848bc95f8a9f7e4273",
          "weight": 4
        },
        {
          "account_hash": "account-hash-f19e3269ae27a44d5debfcd2df0a5b5dd6c31961d335a09dcd5f782ddebb3683",
          "weight": 2
        }
      ],
      "main_purse": "uref-250554daef44c5274f726f2556b62da7dbd6d997d6a93dea925104529db12e6a-007",
      "named_keys": [
        {
          "key": "hash-fc8eaf47329d45c76c4f80ec4e6fdb08e74dbe4ae0391447c5ebd15f1c962476",
          "name": "keys_manager"
        },
        {
          "key": "uref-3141039f23ae092028da3b8222ebb5ae5d8409e745b507903e5f7600fa42aada-007",
          "name": "keys_manager_hash"
        }
      ]
    },
    "api_version": "1.4.4",
    "merkle_proof": "[26654 hex chars]"
  }
}

</details>

> ## Link the Other Accounts and Set Key Weights 

Reviewer: ```suggestion
## Linking the Other Accounts and Setting Key Weights 
```

                },
                {
                  "key": "balance-34b99d631b7545a56af663801f8b86c0f38cc014b002e9f7e0c0b2a19e789f14",
                  "transform": "Identity"
                },
                {
                  "key": "balance-98d945f5324f865243b7c02c0417ab6eac361c5c56602fd42ced834a1ba201b6",
                  "transform": {
                    "WriteCLValue": {
                      "bytes": "00",
                      "cl_type": "U512",
                      "parsed": "0"
                    }
                  }
                },
                {
                  "key": "balance-34b99d631b7545a56af663801f8b86c0f38cc014b002e9f7e0c0b2a19e789f14",
                  "transform": {
                    "AddUInt512": "100000000000"
                  }
                }
              ]
            },
            "transfers": []
          }
        }
      }
    ]
  }
}

</details>

### View account details {#view-account-details}

The following command gets the account details after the deploy is successful:

```
casper-client get-account-info \
--public-key <hexadecimal public key> \
--node-address http://[NODE_IP_ADDRESS]:7777
```

> In the output of this command, you can see the key weight, key management threshold, and deploy threshold. Also, observe the main purse structure, which has the `keys_manager_hash` uref address. This uref address is used to find the smart contract's [session hash](#find-session-hash).

Reviewer 2: ```suggestion
In the output of this command, you can see the key weight, key management threshold, and deploy threshold. Also, observe the main purse structure, which has the `keys_manager_hash` uref address. In the next steps, this uref address is used to find the smart contract's [session hash](#find-session-hash).
```

<details>
<summary>Sample output with named keys for the keys manager contract</summary>

{
  "id": 3805776763924596530,
  "jsonrpc": "2.0",
  "result": {
    "account": {
      "account_hash": "account-hash-694c000a8282d039ace20fbad7a692f4d800a7fd6db08848bc95f8a9f7e4273",
      "action_thresholds": {
        "deployment": 1,
        "key_management": 1
      },
      "associated_keys": [
        {
          "account_hash": "account-hash-694c000a08282d039ace20fbad7a692f4d800a7fd6db08848bc95f8a9f7e4273",
          "weight": 1
        }
      ],
      "main_purse": "uref-250554daef44c5274f726f2556b62da7dbd6d997d6a93dea925104529db12e6a-007",
      "named_keys": [
        {
          "key": "hash-fc8eaf47329d45c76c4f80ec4e6fdb08e74dbe4ae0391447c5ebd15f1c962476",
          "name": "keys_manager"
        },
        {
          "key": "uref-3141039f23ae092028da3b8222ebb5ae5d8409e745b507903e5f7600fa42aada-007",
          "name": "keys_manager_hash"
        }
      ]
    },
    "api_version": "1.4.4",
    "merkle_proof": "[26588 hex chars]"
  }
}

</details>

> In the above output, you can see the named keys for the contract and the uref for keys_manager_hash. This uref hash will be used to find the session hash of the contract.

Reviewer 2: I would consider wraping all of the `keys_manager_hash` and values from client output into `code` blocks
